plugins/UsersManager/API.php

```diff
@@ -780,16 +805,41 @@ private function isUserTheOnlyUserHavingSuperUserAccess($userLogin)
     }
 
     /**
+     * Generates a new random authentication token.
+     *
+     * @param string $userLogin Login
+     * @return string
+     */
+    public function createTokenAuth($userLogin)
+    {
+        return hash(
+            'sha256',
+            $userLogin . microtime(true) . Common::generateUniqId() . SettingsPiwik::getSalt()
+        );
+    }
+
+    /**
      * Generates a unique MD5 for the given login & password
      *
      * @param string $userLogin Login
      * @param string $md5Password hashed string of the password (using current hash function; MD5-named for historical reasons)
      * @return string
+     * @throws Exception if credentials do not match
      */
     public function getTokenAuth($userLogin, $md5Password)
     {
         UsersManager::checkPasswordHash($md5Password, Piwik::translate('UsersManager_ExceptionPasswordMD5HashExpected'));
 
-        return md5($userLogin . $md5Password);
+        $user = $this->model->getUser($userLogin);
+
+        if (!empty($user) && password_verify($md5Password, $user['password'])) {
+            if (password_needs_rehash($user['password'], PASSWORD_BCRYPT)) {
+                $this->model->updateUser($login, $md5Password, $user['alias'], $user['email'], $user['token_auth']);
+            }
+
+            return $user['token_auth'];
+        }
+
+        throw new Exception(Piwik::translate('UsersManager_ExceptionTokenAuthCredentials', $userLogin));
```

Comment:
I'm not 100% sure about this. I think we should maybe in this case still return a random token as existing applications may depend on this behaviour. Eg in Piwik Mobile we call this method, and then check whether the returned token is valid by calling `SitesManager.getSitesWithViewAccess`. We would need to test if we break Piwik Mobile by throwing an exception here. Would it be possible to simple generate a random token here and return it? It should not be based on the Piwik salt that is used in `createTokenAuth` as one could otherwise over time guess the Piwik salt. Maybe just `return hash(
            'sha256',
            $userLogin . Common::generateUniqId())
        );`